google/cloud/sql/connector/InstanceConnectionManager.py

```diff
@@ -204,14 +254,82 @@ def __init__(
 
         data = {"public_key": pub_key}
 
-        resp = await client_session.post(
-            url, headers=headers, json=data, raise_for_status=True
-        )
+        if client_session is not None:
+            resp = await client_session.post(
+                url, headers=headers, json=data, raise_for_status=True
+            )
+        else:
+            async with aiohttp.ClientSession() as cs:
+                resp = await cs.post(
+                    url, headers=headers, json=data, raise_for_status=True
+                )
+
         ret_dict = json.loads(await resp.text())
 
         return ret_dict["cert"]
 
-    def _auth_init(self):
+    async def _get_instance_data(
+        self,
+    ) -> Dict[str, Union[OpenSSL.SSL.Context, Dict]]:
+        """Asynchronous function that takes in the futures for the ephemeral certificate
+        and the instance metadata and generates an OpenSSL context object.
+
+        :rtype: Dict[str, Union[OpenSSL.SSL.Context, Dict]]
+        :returns: A dict containing an OpenSSL context that is created using the requested ephemeral certificate
+            instance metadata and a dict that contains all the instance's IP addresses.
+        """
+
+        print("Creating context")
+
+        metadata_task = self._loop.create_task(
+            self._get_metadata(
+                self._client_session, self._credentials, self._project, self._instance
+            )
+        )
+
+        ephemeral_task = self._loop.create_task(
+            self._get_ephemeral(
+                self._client_session,
+                self._credentials,
+                self._project,
+                self._instance,
+                self._pub_key.decode("UTF-8"),
+            )
+        )
+
+        metadata, ephemeral_cert = await asyncio.gather(metadata_task, ephemeral_task)
+
+        PEM = OpenSSL.crypto.FILETYPE_PEM
+
+        pkey = OpenSSL.crypto.load_privatekey(PEM, self._priv_key)
```

Comment:
`pkey` -> `public_key`

A comment for this section explain what is doing on wouldn't go amiss. (In fact, it might make it easier to read this function if you move this part into it's own function)